### PR TITLE
feat: persist CreDtTm with account creation - #180

### LINKED ISSUE #180: [feat] persist CreDtTm with account creation
## Description
Update the batch-ppa service to pass a `CreDtTm` (creation timestamp) parameter when creating account records, ensuring consistency with real-time transaction processing and maintaining accurate audit trails for batch-processed accounts.

## Background
The account table now includes a `creDtTm` field to capture the creation timestamp. When batch-ppa creates account records during batch processing, it must provide this timestamp to maintain data consistency and audit compliance.

## Investigation Required

⚠️ **First Step: Determine if batch-ppa creates accounts**

Before implementing changes, we need to confirm:
- [ ] Does batch-ppa create new account records?
- [ ] Does it call `saveAccount` or similar methods?
- [ ] Does it use frms-coe-lib database manager?
- [ ] Does it make direct database INSERT operations?

**If NO account creation**: This issue can be closed or converted to a low-impact documentation update.

**If YES account creation**: Proceed with changes below.

### 1. Update Database Manager Calls

**File**: `src/[relevant-service-file].ts` (TBD based on codebase structure)

Update any `saveAccount` calls to include CreDtTm:

**Before**:
```typescript
await databaseManager.saveAccount(accountId, tenantId);
```

**After**:
```typescript
const creDtTm = determineAccountCreationTimestamp(batchRecord);
await databaseManager.saveAccount(accountId, tenantId, creDtTm);
```

### 2. Implement Timestamp Strategy

**File**: `src/utils/timestamp-handler.ts` (new file)

Create a utility to determine the appropriate timestamp for batch-processed accounts:

```typescript
/**
 * Determines the creation timestamp for batch-processed accounts
 * 
 * Priority order:
 * 1. Transaction-level timestamp from batch record
 * 2. Batch file timestamp
 * 3. Current processing time (fallback)
 */
export function determineAccountCreationTimestamp(
  batchRecord: any,
  batchMetadata?: { timestamp?: string }
): string {
  // Option 1: Extract from batch record if available
  if (batchRecord?.CreDtTm) {
    try {
      const date = new Date(batchRecord.CreDtTm);
      if (!isNaN(date.getTime())) {
        return date.toISOString();
      }
    } catch (error) {
      // Invalid timestamp, continue to next option
    }
  }

  // Option 2: Use batch file timestamp
  if (batchMetadata?.timestamp) {
    try {
      const date = new Date(batchMetadata.timestamp);
      if (!isNaN(date.getTime())) {
        return date.toISOString();
      }
    } catch (error) {
      // Invalid timestamp, continue to next option
    }
  }

  // Option 3: Use current processing time as fallback
  return new Date().toISOString();
}
```

### 3. Update Batch Processing Logic

**File**: `src/[batch-processor].ts` (TBD)

Example implementation in batch processing flow:

```typescript
import { determineAccountCreationTimestamp } from './utils/timestamp-handler';

async function processBatchTransaction(
  transaction: any,
  batchMetadata: { timestamp?: string }
): Promise<void> {
  const accountId = extractAccountId(transaction);
  const tenantId = extractTenantId(transaction);
  
  // Determine appropriate timestamp
  const creDtTm = determineAccountCreationTimestamp(transaction, batchMetadata);
  
  // Create account with timestamp
  if (!await accountExists(accountId, tenantId)) {
    await databaseManager.saveAccount(accountId, tenantId, creDtTm);
  }
  
  // Continue with transaction processing...
}
```

### 4. Update Batch File Metadata Handler

**File**: `src/[batch-file-handler].ts` (TBD)

Capture batch file timestamp when loading batch:

```typescript
async function loadBatchFile(filePath: string): Promise<BatchData> {
  const fileStats = await fs.stat(filePath);
  
  const batchMetadata = {
    fileName: path.basename(filePath),
    timestamp: fileStats.mtime.toISOString(),  // File modification time
    fileSize: fileStats.size,
  };
  
  const transactions = await parseBatchFile(filePath);
  
  return {
    metadata: batchMetadata,
    transactions,
  };
}
```

### 5. Update Tests

**File**: `__tests__/unit/batch-processing.test.ts` (TBD)

Update tests to include CreDtTm:

```typescript
describe('Batch Account Creation', () => {
  it('should create account with transaction timestamp', async () => {
    const mockTransaction = {
      accountId: 'ACC123',
      tenantId: 'TENANT1',
      CreDtTm: '2026-02-11T10:00:00.000Z',
    };
    
    const mockDbManager = {
      saveAccount: jest.fn().mockResolvedValue(undefined),
    };
    
    await processTransaction(mockTransaction, {});
    
    expect(mockDbManager.saveAccount).toHaveBeenCalledWith(
      'ACC123',
      'TENANT1',
      '2026-02-11T10:00:00.000Z'
    );
  });
  
  it('should use batch file timestamp when transaction timestamp missing', async () => {
    const mockTransaction = {
      accountId: 'ACC123',
      tenantId: 'TENANT1',
      // No CreDtTm in transaction
    };
    
    const batchMetadata = {
      timestamp: '2026-02-11T09:00:00.000Z',
    };
    
    await processTransaction(mockTransaction, batchMetadata);
    
    expect(mockDbManager.saveAccount).toHaveBeenCalledWith(
      'ACC123',
      'TENANT1',
      '2026-02-11T09:00:00.000Z'
    );
  });
  
  it('should use current time as fallback', async () => {
    const mockTransaction = {
      accountId: 'ACC123',
      tenantId: 'TENANT1',
    };
    
    const beforeTime = new Date().toISOString();
    await processTransaction(mockTransaction, {});
    const afterTime = new Date().toISOString();
    
    const actualTimestamp = mockDbManager.saveAccount.mock.calls[0][2];
    expect(actualTimestamp).toBeGreaterThanOrEqual(beforeTime);
    expect(actualTimestamp).toBeLessThanOrEqual(afterTime);
  });
  
  it('should handle invalid timestamp gracefully', async () => {
    const mockTransaction = {
      accountId: 'ACC123',
      tenantId: 'TENANT1',
      CreDtTm: 'INVALID-DATE',
    };
    
    // Should not throw, should use fallback
    await expect(processTransaction(mockTransaction, {})).resolves.not.toThrow();
  });
});
```

### 6. Update Configuration/Documentation

**File**: `README.md`

Add section documenting timestamp behavior:

```markdown
## Batch Account Creation Timestamps

When batch-ppa creates account records, it determines the `creDtTm` timestamp using the following priority:

1. **Transaction-level timestamp**: If the batch transaction includes a `CreDtTm` field
2. **Batch file timestamp**: The file modification time when the batch was uploaded
3. **Processing time**: Current system time as fallback

### Timestamp Format

- Format: ISO 8601 (e.g., `2026-02-11T14:30:45.123Z`)
- Timezone: Always UTC
- Precision: Milliseconds

### Configuration

Set batch file timestamp behavior via environment variable:

```env
# Use file modification time as batch timestamp
BATCH_TIMESTAMP_SOURCE=file_mtime

# Or use file creation time
BATCH_TIMESTAMP_SOURCE=file_ctime

# Or use current processing time
BATCH_TIMESTAMP_SOURCE=processing_time
```
```

## Timestamp Strategy Decision

⚠️ **Decision Required**: Choose timestamp strategy for batch processing

| Strategy | Pros | Cons | Recommended For |
|----------|------|------|-----------------|
| **Transaction-level** | Most accurate, preserves original timestamp | Requires batch format to include timestamps | Batches with per-transaction timestamps |
| **Batch file** | Represents batch creation time, simple | Less granular than transaction-level | Historical batch imports |
| **Processing time** | Simple implementation, no dependencies | Least accurate, loses historical context | Low-priority batches |

**Recommendation**: Use transaction-level if available, fallback to batch file timestamp, with processing time as last resort (as implemented in the utility above).

## Integration Considerations

### Batch vs Real-Time Timestamps

**Scenario**: Account created in batch, then appears in real-time transaction

```typescript
// Batch processing (earlier)
await saveAccount('ACC123', 'TENANT1', '2026-02-01T10:00:00Z');

// Real-time processing (later) - should not update timestamp
await saveAccount('ACC123', 'TENANT1', '2026-02-11T15:00:00Z');
// ON CONFLICT DO NOTHING - preserves original timestamp ✓
```

The `ON CONFLICT DO NOTHING` strategy ensures the first timestamp (typically from batch) is preserved.

### Batch Reprocessing

If batches are reprocessed:
- Existing accounts retain their original `creDtTm`
- New accounts in reprocessed batch get reprocessing timestamp (or batch file timestamp)
- Consider logging batch reprocessing events separately

## Dependencies

**Blocking Dependencies:**
- [ ] frms-coe-lib: Update `saveAccount` signature to accept CreDtTm
- [ ] full-stack-docker-tazama: Add `creDtTm` column to account table

**Related Issues:**
- frms-coe-lib: Add CreDtTm parameter to saveAccount
- full-stack-docker-tazama: Add creDtTm column to account table
- tms-service: Pass CreDtTm to saveAccount
- admin-service: Pass CreDtTm when creating accounts

## Acceptance Criteria

**Investigation Phase:**
- [ ] Confirmed whether batch-ppa creates accounts
- [ ] Identified all account creation locations in codebase
- [ ] Determined batch file format and available timestamp fields
- [ ] Chosen timestamp strategy (transaction/batch-file/processing-time)

**Implementation Phase (if accounts are created):**
- [ ] Updated all `saveAccount` calls to include CreDtTm
- [ ] Implemented timestamp determination utility
- [ ] Added timestamp extraction from batch records
- [ ] Added timestamp extraction from batch file metadata
- [ ] Implemented fallback to processing time
- [ ] Updated unit tests for account creation
- [ ] Updated integration tests for batch processing
- [ ] Added tests for timestamp priority logic
- [ ] Added tests for invalid timestamp handling
- [ ] Updated documentation with timestamp behavior
- [ ] Added configuration for timestamp strategy

**Testing Phase:**
- [ ] Test with batches containing transaction timestamps
- [ ] Test with batches without transaction timestamps
- [ ] Test with invalid timestamp formats
- [ ] Test timestamp fallback logic
- [ ] Test batch reprocessing scenarios
- [ ] Test account creation in batch vs real-time
- [ ] Performance test with large batch files
- [ ] Verify timestamps stored correctly in database

## Estimated Effort

**Investigation**: 0.5 days
- Identify account creation locations
- Review batch file formats
- Determine timestamp strategy

**Implementation** (if accounts are created): 2-3 days
- Update saveAccount calls
- Implement timestamp utilities
- Update batch processing logic

**Testing**: 1-2 days
- Unit tests
- Integration tests
- Edge case testing

**Documentation**: 0.5 days

**Total**: 4-6 days (if account creation confirmed)

### Batch File Scenarios
- [ ] Batch with transaction-level timestamps
- [ ] Batch without transaction timestamps
- [ ] Batch with invalid timestamp formats
- [ ] Batch with mixed timestamp availability
- [ ] Empty batch file
- [ ] Corrupted batch file

### Timestamp Scenarios
- [ ] Valid ISO 8601 timestamps
- [ ] Invalid date strings
- [ ] Future timestamps (allowed)
- [ ] Very old timestamps (allowed)
- [ ] Missing timestamps (fallback)

### Integration Scenarios
- [ ] Account created in batch, then seen in real-time
- [ ] Account created in real-time, then seen in batch
- [ ] Batch reprocessing with existing accounts
- [ ] Multiple batches creating same account

### Edge Cases
- [ ] System clock issues
- [ ] Timezone handling
- [ ] Daylight saving time transitions
- [ ] Leap seconds (if supported)

## Risk Assessment

| Risk | Likelihood | Impact | Mitigation |
|------|------------|--------|------------|
| Batch format lacks timestamps | Medium | Medium | Implement robust fallback strategy |
| Large batch processing slowdown | Low | Medium | Optimize timestamp extraction |
| Timestamp timezone confusion | Medium | High | Always use UTC, validate format |
| Historical data accuracy | Medium | Medium | Document timestamp source clearly |

## Notes

- Ensure batch processing logs include timestamp source (transaction/batch-file/processing-time) for audit purposes
- Consider adding a separate field to track "batch processing time" vs "account creation time"
- Monitor for timestamp anomalies (e.g., all accounts in batch having identical timestamps)
- Consider rate limiting or batching database operations for performance

## Open Questions

1. What is the typical batch file format? (CSV, JSON, XML, proprietary?)
2. Do batch files include per-transaction timestamps?
3. What is the typical batch size? (affects performance considerations)
4. Are batches ever reprocessed? (affects timestamp handling)
5. Should we distinguish between "batch uploaded time" vs "account creation time"?
6. Is there a requirement to preserve original timestamps from source systems?

### PR DESCRIPTION
**Pull request by abdul-rahim-psl**
# SPDX-License-Identifier: Apache-2.0

## What Did We Change?

src/utils/timestamp-handler.ts - Created timestamp utility with 3-tier fallback
src/services/save.transactions.service.ts - Integrated timestamp utility into all transaction handlers
src/services/file.service.ts - Added batch metadata extraction from files
src/utils/helper.functions.ts - Added batch metadata parameter
__tests__/unit/batch-processing.test.ts - Added 4 comprehensive tests

**Core Changes:**
Timestamp Priority: Transaction timestamp → File timestamp → Current time
Account Creation: All addAccount() calls now include CreDtTm parameter
Metadata Support: Batch file stats (modification time, size, name) passed through processing chain
Code Quality: Fixed ESLint errors (array destructuring, magic numbers, type safety)

## Why Are We Doing This?

Account table requires creDtTm field for audit compliance.

Regulatory Requirement: Financial monitoring systems must track account creation timestamps
Data Consistency: Batch-processed accounts need same audit trail as real-time transactions
Compliance: Maintains regulatory standards for transaction monitoring
Result: Batch processing now meets audit requirements while preserving system reliability.

## How was it tested?
- [x] Locally
- [x] Development Environment
- [ ] Not needed, changes very basic
- [x] Husky successfully run
- [x] Unit tests passing and Documentation done

## Summary by CodeRabbit

* **New Features**
  * Improved account-creation timestamp handling: normalized ISO timestamps, graceful fallbacks to batch or current time, and propagation of optional batch/file metadata through processing.

* **Tests**
  * Added unit tests covering timestamp selection, fallbacks, and invalid/missing timestamp scenarios.

* **Chores**
  * Updated a library dependency version.

### CODE REVIEW

File: src/utils/timestamp-handler.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * 3. Current processing time (fallback)
+ */
+export function determineAccountCreationTimestamp(batchRecord: CreationTimeSource, batchMetadata?: { timestamp?: string }): string {
+  // Option 1: Extract from batch record if available
+  if (batchRecord.CreDtTm) {
+    try {
+      const date = new Date(batchRecord.CreDtTm);
+      if (!isNaN(date.getTime())) {
+        return date.toISOString();
+      }
```
[the remark this answers is not in the document]
Comment (abdul-rahim-psl) [reply src/utils/timestamp-handler.ts]: done. thanks

**Review by MAdeel95 — DISMISSED** [review]

File: src/services/save.transactions.service.ts [hunk 2]
```diff
@@ -4,30 +4,45 @@ import type { DataCache, Pacs008, Pain001, Pain013, TransactionDetails } from '@
...
+  transactionType: string,
+  batchMetadata?: BatchMetadata,
+): Promise<Pain001 | boolean> => {
   const id = transaction.CstmrCdtTrfInitn.GrpHdr.MsgId;
   loggerService.trace('Start - Handle transaction data', 'handlePain001()', id);
   const span = apm.startSpan('transaction.pain001');
   const TxTp = transactionType;
   transaction.TxTp = TxTp;
-  const TenantId = transaction.TenantId ?? 'DEFAULT';
+  const TenantId = transaction.TenantId || 'DEFAULT';
```
[the remark this answers is not in the document]
Comment (abdul-rahim-psl) [reply src/services/save.transactions.service.ts]: done. thanks

**Review by MAdeel95 — APPROVED** [review]

**Review by MAdeel95 — DISMISSED** [review]
